File: backend/app/api/v1/agents.py
"""
Agent 管理 API - 完整的 CRUD + 状态管理
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.models.user import User
from app.schemas.agent import (
    AgentCreate,
    AgentUpdate,
    AgentResponse,
    AgentListResponse,
    AgentStatusUpdate,
)
from app.services.auth_service import get_current_user
from app.services import agent_service

router = APIRouter()


@router.get("/", response_model=AgentListResponse)
async def list_agents(
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    status: str = Query(None, description="按状态筛选"),
    search: str = Query(None, description="搜索关键词"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """获取 Agent 列表（分页 + 筛选）"""
    agents, total = await agent_service.list_agents(
        db=db,
        page=page,
        page_size=page_size,
        status=status,
        search=search,
        created_by=current_user.id,
    )
    return AgentListResponse(
        items=[AgentResponse.model_validate(a) for a in agents],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("/", response_model=AgentResponse, status_code=201)
async def create_agent(
    data: AgentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """创建新的 Agent"""
    # 如果没有指定 workspace_id，使用默认值
    if not data.workspace_id:
        data.workspace_id = f"default_{current_user.id}"

    agent = await agent_service.create_agent(db, data, current_user.id)
    return AgentResponse.model_validate(agent)


@router.get("/{agent_id}", response_model=AgentResponse)
async def get_agent(
    agent_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """获取 Agent 详情"""
    agent = await agent_service.get_agent(db, agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return AgentResponse.model_validate(agent)


@router.put("/{agent_id}", response_model=AgentResponse)
async def update_agent(
    agent_id: str,
    data: AgentUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """更新 Agent 配置"""
    agent = await agent_service.update_agent(db, agent_id, data)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return AgentResponse.model_validate(agent)


@router.delete("/{agent_id}", status_code=204)
async def delete_agent(
    agent_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """删除 Agent"""
    success = await agent_service.delete_agent(db, agent_id)
    if not success:
        raise HTTPException(status_code=404, detail="Agent not found")
    return None


@router.patch("/{agent_id}/status", response_model=AgentResponse)
async def update_agent_status(
    agent_id: str,
    data: AgentStatusUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """变更 Agent 状态（draft/running/stopped/error/archived）"""
    try:
        agent = await agent_service.update_agent_status(db, agent_id, data.status)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        return AgentResponse.model_validate(agent)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/api/agents.ts
import client from './client'

export interface AgentInfo {
  id: string
  name: string
  description: string | null
  avatar: string | null
  status: string
  model_provider: string | null
  model_name: string | null
  model_config_template_id: string | null
  temperature: number | null
  max_tokens: number | null
  context_window: number | null
  system_prompt: string | null
  welcome_message: string | null
  enabled_skills: string | null
  enabled_mcp_servers: string | null
  workspace_id: string | null
  created_by: string
  created_at: string
  updated_at: string | null
}

export interface AgentListResponse {
  items: AgentInfo[]
  total: number
  page: number
  page_size: number
}

export interface AgentCreatePayload {
  name: string
  description?: string | null
  avatar?: string | null
  system_prompt?: string | null
  welcome_message?: string | null
  model_provider?: string | null
  model_name?: string | null
  model_config_template_id?: string | null
  temperature?: number | null
  max_tokens?: number | null
  context_window?: number | null
  enabled_skills?: string[] | null
  enabled_mcp_servers?: string[] | null
  workspace_id?: string | null
  status?: string
}

export interface AgentUpdatePayload {
  name?: string
  description?: string | null
  avatar?: string | null
  system_prompt?: string | null
  welcome_message?: string | null
  model_provider?: string | null
  model_name?: string | null
  model_config_template_id?: string | null
  temperature?: number | null
  max_tokens?: number | null
  context_window?: number | null
  enabled_skills?: string[] | null
  enabled_mcp_servers?: string[] | null
}

/** 获取 Agent 列表（分页 + 筛选） */
export async function listAgents(params?: {
  page?: number
  page_size?: number
  status?: string
  search?: string
}) {
  const { data } = await client.get<AgentListResponse>('/agents/', { params })
  return data
}

/** 获取单个 Agent */
export async function getAgent(id: string) {
  const { data } = await client.get<AgentInfo>(`/agents/${id}`)
  return data
}

/** 创建 Agent */
export async function createAgent(payload: AgentCreatePayload) {
  const { data } = await client.post<AgentInfo>('/agents/', payload)
  return data
}

/** 更新 Agent */
export async function updateAgent(id: string, payload: AgentUpdatePayload) {
  const { data } = await client.put<AgentInfo>(`/agents/${id}`, payload)
  return data
}

/** 删除 Agent */
export async function deleteAgent(id: string) {
  await client.delete(`/agents/${id}`)
}

/** 更新 Agent 状态 */
export async function updateAgentStatus(id: string, status: string) {
  const { data } = await client.patch<AgentInfo>(`/agents/${id}/status`, {
    status,
  })
  return data
}

/* ── 本地 Agent 发现 ──────────────────────────────────── */

export interface DiscoveredModel {
  source: string
  source_name: string
  model_name: string
  provider: string
  size?: number
  endpoint: string
}

export interface DiscoverResponse {
  items: DiscoveredModel[]
  total: number
  message: string
}

/** 扫描本地 AI 服务，发现可用模型 */
export async function discoverAgents() {
  const { data } = await client.post<DiscoverResponse>('/discover/agents')
  return data
}

/** 将发现的模型注册为 Agent */
export async function registerDiscoveredAgent(params: {
  model_name: string
  provider?: string
  endpoint?: string
  workspace_id?: string
}) {
  const { data } = await client.post('/discover/agents/register', null, { params })
  return data
}
